### PR TITLE
Adds Object3D indices referencing and 'replace' object/index

### PR DESCRIPTION
**Pull request by MasterJames**
This adds index functionality to the children manipulators of Object3D.
'add' calls 'addAt' but addAt is advanced and can also reorder emitting a new 'moved' event.
'remove' is able to use an index or object and since it was more efficient to use splice you can include item to insert as well, making replace and replaceAt just aliases to the mighty quasi-remove function, which handles inserting thanks to splice.

### CODE REVIEW

**Comment by MasterJames** [comment]
Sadly this still has reformatting issues.

File: src/core/Object3D.js [hunk 1]
```diff
@@ -91,628 +91,682 @@ THREE.Object3D.DefaultMatrixAutoUpdate = true;
...
+    add: function () {
 
-		if ( arguments.length > 1 ) {
+        var i = 0, l = arguments.length;
 
-			for ( var i = 0; i < arguments.length; i ++ ) {
+        for ( ; i < l ; i ++ ) {
 
-				this.add( arguments[ i ] );
+            this.addAt( undefined, arguments[ i ] );
```
Comment (mrdoob) [inline src/core/Object3D.js]: I would just do:

```javascript
this.addAt( this.children.length, arguments[ i ] );
```

**Comment by EliasHasle** [comment]
> Sadly this still has reformatting issues.
@MasterJames That is really hard to read. Couldn't you just do the changes again in the internal editor to avoid the duplication thing? Considering that this changes core parts of THREE, proper proofreading is important for many people, and hopefully it is a requirement for these changes to be accepted.

I appreciate the effort to approach my original request, but am a bit skeptical against making a lot of changes to the current methods. In particular, making add use addAt as a helper may spare some lines of code, but the extra calls and checking may incur a computational/memory cost when one calls many add(object). I have similar concerns about some other changes, admitting that I haven't digged through the entire diff.

Adding replace and replaceAt first seems much safer, less invasive.

**Comment by MasterJames** [comment]
This PR is bogus. I'm redoing it sorry for rushing it.

**Comment by MasterJames** [comment]
Right now remove works better with no labeled arguments. Before it calls itself checking if args is greater then 1 repeatedly. Remove takes a list fine still. Replace can be either an object or index and removeAt is added for user clarify but it just calls replace without a replacement.
Event Dispatch is optimized with %50 loop deletion (probably shold be a separate PR).

**Comment by EliasHasle** [comment]
Regarding EventDispatcher: I would say that the original code looks so "silly" at first impression that it must have some non-obvious purpose. Is there a possibility that the array copied from is somehow considered volatile, and that copying it before starting to do calls based on it, avoids some concurrency errors, because then the array looped over isn't changed while in the loop? (If so, testing it without massive event concurrency is not sufficient to detect the difference.)

**Comment by MasterJames** [comment]
It is the initial assumption upon seeing it, but I thought there is an equal reverse argument that in the same way not calling instantly could also result in it being freed etc.
If this was in the suggested reduced form, what would would be needed on the other end to avoid the problem?
It also suggests as much as I hate the idea but is there a way to test this sort of thing at all or is it best to live in fear or make an assumption its a problem with no alternative solution?
Which examples might do child creation and destruction/removal (and ideally event listening) as well I wondered?
I'm coining the phase Event Mistening, just kidding.

**Comment by MasterJames** [comment]
Ah maybe it should not move if it's setting index to the end by default from 'add'.

**Comment by MasterJames** [comment]
I'm currently doing QUnit tests to make sure this is working perfectly.
I think addAt will have an optional move ability while replace won't. 
The concern there is if you allow move and target an index (above the existing object to be moved)  there is a shift of indices while moving so if it's above it will be a lower number. So in a way what you also then need to know is, "Is that the index  before or the index number after it extracts it if it finds it and the new location is positioned after the moved source's index. {I think you get my drift}.
Still when you have it optional if it's an object too and not just an index, well then I think this makes more sense, that you're not as concerned with index just swapping items with the same code.
[Swapping also came to mind.]
Anyway I'll make a bunch of QUnit tests for all these options and abilities and so that way we'll know if it's working well etc.

**Comment by MasterJames** [comment]
Okay I believe this is ready for another look, please and thank you.
I think parts of this are exciting and well done.
It was much more challenging then at first was anticipated. The complexity went up a little bit for the sake of completeness. This is how I've solved the problem and refactored things a little.
I hope to learn many things from all the great input this could use.

I added QUnit tests and refactored things a few times.
removeAll calls remove with a list of all indices
remove calls replace without anything to replace
add calls addAt with all arguments passed and no target insertion point
addAt takes an entity to search for the location to insert the Object [second argument]
and adds arguments to allow 'move' and a 'shift' to control how to adjust the target if it's an index and being moved from below.
replace works like addAt in searching but replaces and never moves if it already exists.
findChild is key to acquiring the targets. It will search for a match starting first as before for an Object3D, then it tries for a index, before moving on to a searchOrder array that is used to check for properties on the object it's default is 'name', and 'uuid'. {Warning don't use 'id' there it will conflict with indices}.
The QUnit tests could be improved upon clarified somewhat but a larger inclusive sweep of random changes seemed a better approach initially.

I'm also thinking it needs a sweep of Event Qunit tests, but I was planning to move onto that for EventDispatcher to help prove of disprove the suggested optimization there.
This would  also confirm that using these new indexing tools Events are firing properly and none missed on-mass to make sure 'move' and 'replace' type events are happening properly too.

**Comment by EliasHasle** [comment]
Well, you sure have done a lot of work. Kudos for that. :-)
And I am happy to see that you have split out the change in EventDispatcher.

Then again, you asked for "tough love"... :
I struggle to understand the code. I think the primary reason for this is the usage of very complicated helper functions. (One secondary reason is unclear variable names.) I notice that there are some cases of unnecessary search operations, that will make this code slower than the current version for some common operations. I wonder a bit why you want to use the complicated helper functions for operations that can be accomplished without doing all the tests and searches. There is a real risk that you are introducing bugs in very important functions by doing that kind of abstractions.

I can only speak for myself, but I clearly prefer more transparent procedures that do their stated job simply and efficiently. If I were admin of an open-source project, I would rather deploy untested code that was easy to understand, than thorougly tested code that was hard to understand. Even if I understood the code myself. Because more transparent code invites more community participation, and because there are always some bugs or unintended consequences anyway, and they are easier to find in transparent code.

I will try to update my own "competing" pull request soon. It is intended as a preliminary, very careful introduction of a few features that I believe are useful, and that have been discussed a btit beforehand. I hope I can make it clear and non-invasive enough that it is accepted. I am not "running in the race" of adding a lot of features, and I wish you luck with your bigger project of doing that. I believe I have pointed you in the right direction.

**Comment by EliasHasle** [comment]
As an example of what I mean, here is a sketch for an alternative formulation of removeAll:
function () {
    for(var i=0; i<this.children.length; i++) {
        this.children[i].parent = null;
        this.children[i].dispatchEvent({type: "removed"});
    }
    this.children = [];
}

No abstraction, it just does what it is supposed to, in a simple, efficient way. I haven't tested it, and am not certain that the "cheating" tih dipatching the event before removing the child has no unintended consequences. If it does, I can just change it to:
`
function () {
    var child;
    for(var i=0; i<this.children.length; i++) {
        child = this.children[i];
        child.parent = null;
        this.children[i] = undefined;
        this.children[i].dispatchEvent({type: "removed"});
    }
    this.children = [];
}
`

Or, if the events trigger concurrent operations, or to avoid leaving behind undefined entries, or **anyway**, maybe I can change the whole thing into this:
`
function () {
    var child;
    while (this.children.length > 0) {
        child = this.children.pop();
        child.parent = null;
        child.dispatchEvent({type: "removed"});
    }
}
`
See? I just spat out three successive versions of the same method, and none of them invoked the "allmighty mother of all functions". Less debugging and case testing needed. :-)

**Comment by EliasHasle** [comment]
By the way, shouldn't #8388 and #8389 be closed? :-)

**Comment by MasterJames** [comment]
Well somewhat obviously as the complexity and reuse goes Up its better to consolidate to a findChild function (then vaired duplicates).
[ I think if a reference is used only twice it's not worth the cycles to store it again local in the function etc. to save a second lookup, but 3 times you should extract and reference. There are pros and cons to both approaches but indexing seemed to add more complexity to the problem.
For me complexity is not as much a problem but then I wrote it. Certainly "its too hard to understand" can't override the advantages.
I think if you build your version to match these tests and possible add some more and still achieve as much future flexibility as added property searches does, great!
I always see the 'name' property empty do we really use UUIDs?  Maybe we would more if such access was available and examples given.
I would add both object and index search to the new search list so there's exact control then you can set it to Object3D only or in your case index first/only, which by the way would likely be more efficient but since it searches for objects first there's no real impact vs the previous version.
The core object is better for more people with more functionality. I think the more you think about it the more you'll agree with it.
It would be nice if you tested it with your situation.
I know we are adding unrequested features but the use cases are clearly imaginable and better. Find by UUIDs? 

Anyway it's clear to me Events need stress tests and enough to validate the need for the original EventDispatch I moved to a separate discussion although it's related to this for sure but also to confirm this other part is working correctly.
A final point having replace and not also replaceAt early on with the lookup Entity logic is a smart simplification that avoids confusion in the long run.
There's some pretty smart compactifications in this version too so I think it's pretty good.
Adding children in general and adding index referencing is not going to impact the speed of you use case or many others and those that do can always access children directly so it's just really opened up wide (almost all the way) how to find a match.
I've integrated regular expressions and other stuff differently already but maybe I would be able to integrate this with that idea.
Finally it's maybe like I originally suggested just interject the replaceAt you want via your own require file with 
THREE.Object3D.prototype.replace = function () {}
Otherwise it's complex enough of a problem in my opinion to have to make a real change via a PR with breadth.

**Comment by MasterJames** [comment]
So anyway what I'm saying is "Now with property searching you can just do..."

replace('Hat', newHat );

it's totally trans-formative to the index system anyway.

**Comment by EliasHasle** [comment]
@MasterJames I think that specifically is a cool feature. I have been thinking along similar paths, even with children being used as a hash table, rather than as an array. But one can't change everything at once... And I guess arrays are more optimized in browsers.

**Comment by MasterJames** [comment]
That's true, but you can put index values as keys to objects so it's not impossible to unify that way too. It is what I normally try to do, if any advantage is there for the taking.

**Comment by EliasHasle** [comment]
replace('Hat', newHat );

interpreted as a _set method_ (no need to conserve order) can be implemented with the basic safe methods that are already in THREE.Object3D:
parent.remove(parent.getObjectByName('Hat')); //two passes, one splice
parent.add(newHat)

My original idea was to add a safe method for assigning a value to an array element by index, as an optimisation measure when that index is known.

**Comment by MasterJames** [comment]
I've have decided this is the final way [for this PR], (or the rollback if preferred), but this is my recommendation forward. I think it's stands as a 'good idea', and a versatile yet subtly compromising solution that can also improve performance when using different "ref-styles" together, but being more open-ended only leads to interesting possibilities, (as recommended early adding regex would be another idea that builds on this next).
It's been suggested that maybe getObjectByName: should redirected to findChild: as a final option for optimization that remains. I'll look more into that. 
Anyway I think the tests prove it's working for an number of situations, I'm likely leaving them for cleanup if this remains an idle 'good-idea' PR. I've got to add another test or two maybe as well.

With that said I've been planning to separately build tests for Event Dispatch (at scale) with these ideas [as well as for addressing the "double-loop" in Dispatch quandary (simple "to do or not to do" tests, that prove the functionality, with a timing comparison too)].
I should build both tests that go with that question separately (without these changes), and also add event monitoring to these new ideas of 'replace' and 'move', etc. These test should exist to validate before any idea of a commitment to it be considered, because it is a core level change). Both sets will help each other to validate all this works well or not, even if they lack refinement etc. Noted is that tests will also provide protection well into the future, which is a nice bonus especially if it's currently lacking some robust Event Dispatch tests.

**Comment by MasterJames** [comment]
No your right I didn't look into getObjectByProperty before now. Ohps. I think that changes my thinking already.

**Comment by MasterJames** [comment]
Okay I totally refactored in the getObjectByProperty, which allows the same behavior but searching all objects and descendants before moving to the next search order type while the previous one did each type per object., so maybe better that way anyway.

**Comment by mrdoob** [comment]
Hmm, seems like this PR got messed up with the SpotLight PR...

**Comment by MasterJames** [comment]
Well it's at this point I figured no one seemed to care anymore.
https://github.com/mrdoob/three.js/tree/ab70a57ea46c4438a39ab59f9c6ae479b7ad8ffc
And needed to rebase etc. I guess it should be closed. 
**If you'd like I could try to make it able to be merged still somehow?**
I think the original feature requester for this realized using index was not the best option anymore so there's no real use-case for it anymore. It's an interesting approach using findObject and I thought a good go of the idea.
Sorry I'm still orienting on best practices for keeping multiple PRs functional. Any advice or link to information here would be greatly appreciated and clearly mutually beneficial.

**Comment by MasterJames** [comment]
If I close this main fork of dev called dev will it kill my other patch files that I've submitted as PRs?
Should I keep this Main Fork rebased and try to recreate the contents of this indexing PR into a new patch?

**Comment by EliasHasle** [comment]
>I think the original feature requester for this realized using index was not the best option anymore so there's no real use-case for it anymore.
Being the original feature requester, I dispute this assumption. I am still using indices, for reaons outlined in the updated first post of my feature request. But my use-case is limited to replaceAt.

>Well it's at this point I figured no one seemed to care anymore.
Sadly, yes. Or maybe we are just impatient. I guess @mrdoob has a lot of more important tasks, even just in this project. Really interesting stuff going on in other parts of the code. I am awaiting feedback before working more on my own PR.

I think you should close the two outdated PRs, to make the job easier for the boss. Closing the most updated PR would be a waste, right? Cleaning it up is good, of course. :-)

**Comment by MasterJames** [comment]
Hello again Elias. I made the fork from dev called dev submitted it as a PR since I was using git in my editor and github desktop.
Anyway later I added a few files for "examples" and github made them "patch"es under my fork which one can submit as a PR.
My concern is if I close this main fork's PR will it close all the paches too?  
Eventually if I delete the fork to start over will that delete all the patches too?  So then you wait until all the dust settles and delete it all?  Or what?  
I can't seem to start a new fork?  I figure I could just try to extract that previous state into another patch and PR anyway. Close the PR from the fork's main branch. ? I'm not 100% sure what to do any suggestions or links to understanding this part of gittin' my way through life would be appreciated.

**Comment by EliasHasle** [comment]
Hello James (or Master, whichever you prefer). Oh, now I see. So that's why you didn't close the old PRs, then. I would gladly help you if I could, but am a github newbie myself.

**Comment by MasterJames** [comment]
Okay thanks for the hint to close these off.
This recovered PR/patch contains the accumulation of these efforts.
https://github.com/mrdoob/three.js/pull/8611

I think I've sorted out the PR issues, sufficiently (it can get complicated). The history link has a New PR button right there so it's easy in one way but I needed to check there wasn't conflicts in a mental/manual way before proceeding.
For now I'm going to never do a PR from my main fork's branch, and try to work with the branches within, what GitHub labels "Patches".
When it's safe to delete or close the main forks PR is something I will wait on for a bit unless I hear that's not going to close all the patches as PRs too? I'm 99% sure it won't and so need to look that up somewhere somehow still, or wait until those patches are not open.

**Comment by MasterJames** [comment]
Well I finally got my main dev fork purified of those horrible new line issues and all changes so I can rebase and work with the recommend patch route.
https://github.com/mrdoob/three.js/issues/8632

**Comment by tschw** [comment]
Might be OT, but seeing all the fighting with git:

Since the commits in PRs get squashed now, automatic rebase of feature branches that depend on one another no longer works. If PR A includes commits of PR B and PR B gets merged, you usually want to rebase PR A back to `dev`. In this case, you get conflicts were you normally wouldn't.

What worked for me was to do a `reabse -i dev` and remove the already-merged commits (which are apparently not matched to the squash-merge) manually.

**Comment by MasterJames** [comment]
Thanks for the tip. I'm starting to get more familiar thanks for being patient. I think the editor which has VCS Git stuff built in to be a convenience may have made things more complicated and confusing in these regards (Editor [WS9], Git, Github Desktop, GitHub Website, just makes everything more confusing. Still the newline characters are being interpreted/transformed by the editor too, based on autocrlf and it's own reasoning/settings maybe (then there's repo configs and the default git config). Copying from github Raw view, was also not being newline type aware I don't think. I would use the editor (WS9) to check against the main dev branch and see conflicts with line endings only, change them, but committing them saw nothing wrong and so did nothing/failed, so I was getting stuck with conflicts on my end.
I think your suppose to on windows with git use autocrlf to convert everything and when you commit it removes those stupid Carriage Returns it adds (for text files only). Anyway it was a battle of confusion.
I'm still scared if I close this PR it will destroy my other branch PRs somehow, because they were maybe dependent on this. I doubt it so I'll close this now and pick up the pieces if I have to. Anyway some of Git is confusing for sure and this newline glitch I was not expecting. I usually don't use git as I'm an indie dev and work alone, [BTW this also allows for more radical coding style].
The battle showed me more about how the editor I'm using can switch and create branches which I was doing more in Git Desktop so those files where changing but maybe it didn't know what branch I was on in the editor? So still wrapping my head around the best practices for gittin' around with the THREE Repo and PRs.

**Comment by mrdoob** [comment]
> Since the commits in PRs get squashed now, automatic rebase of feature branches that depend on one another no longer works.

Ohh. That's why...

**Comment by tschw** [comment]
@MasterJames 
> I'm still scared if I close this PR it will destroy my other branch PRs somehow, because they were maybe dependent on this.

**Closing the PR is certainly safe. You can even delete the branch. What's in one branch never affects the other. A branch is just a pointer, really.** One commit showing up merged when a dependent one got merged is something Github's UI figures out when it scans whether your commit can be merged (probably doesn't happen anymore, since commits are being squashed).
**You can change your branch to refer to another commit id (which describes a complete, immutable state of the entire working copy) with `git reset`. This does not touch your files. It *does* when you add `--hard` (so careful with that one).**

Also, git doesn't forget commits easily. The ones that are older than 30 days (default) AND not referenced by any branch are removed when you run `git gc` (garbage collect). So **as long as you remember the IDs in the log, you can still restore your branch.** One way to make sure you don't forget them is `git branch MyTempBackupBranch` , saving the log to an untracked text file is another.

Here is an example, rebuilding your changes on current dev:
```shell
$ git checkout dev # go to my dev branch
$ git pull upstream dev # pull what got merged upstream
$ git checkout MyFeature # go to my feature branch
$ git merge dev # bring my feature up to date
$ git reset dev # set branch to last commit in dev without changing my files
$ # note, I can still reset back to my old history when I know the commit id
$ git add src/SomeFile examples/AnotherFile # re-add my feature
$ git commit # commits go on top of dev
$ # could make multiple commits here, of course 
$ git push --force origin MyFeature # rewrite the PR
```

Then there's `git cherry-pick` that lets you pull a commit from another branch, treating it as a diff against its predecessor. This means it gets "diffed-out" of its history and the patch is committed (with a new ID) onto the current branch.

Knowing these basic building blocks, `git rebase dev` is pretty much just an automation to save you a lot of typing. It rewrites a new state of your branch on top of `dev` and cherry-picks all your changes over in the old state, this way keeping your history. It also automatically figures out which commits in your branch got merged already (which doesn't work when these commits were squashed into a single, different commit - so have to help it there). 

> I think your suppose to on windows with git use autocrlf to convert everything and when you commit it removes those stupid Carriage Returns it adds (for text files only). Anyway it was a battle of confusion.

Sounds like you got it all figured out. I once had a similar fight on Windows with some GUI-packaged SCM (not git - don't remember but I think was SVN) and ended up with a similar conclusion. Turned on by default, often just works, but sometimes doesn't - and in this case makes quite a mess.

> Anyway some of Git is confusing for sure [...] I usually don't use git as I'm an indie dev and work alone

Git left me totally puzzled when I started using it. Not fast-forward?! What it want with my stereo?! Told me to "use reset" and I typed `git reset` and absolutely *nothing* happened. Strange and mysterious piece of software...

> BTW this also allows for more radical coding style

Since I got the hang of it a bit, I must say it adapts quite nicely to pretty much anything - no matter how crazy.

@mrdoob 
> > Since the commits in PRs get squashed now, automatic rebase of feature branches that depend on one another no longer works.

> Ohh. That's why...

Yeah, I first noticed that `branch -d` didn't work anymore (had to use forced delete `-D` all the sudden). Getting a conflict rewinding my branch in a rebase, I finally figured out what was going on.

I still find squashed commits good, since we otherwise have tons of commits from edits that are actually unessential to the overall history of the lib (I occasionally "tentacle out" in feature branches, and in these cases it would make sense to keep the history - those commits I make for ease of review would still good to be squashed, so I personally wouldn't mind if you just told me to squash the edits once you're otherwise ready to merge, but that's not a scalable solution, of course).

I've just been looking for options that may allow squashing while adding the refs to the commit message, so that git can figure it out (it does something like that in other cases), but unfortunately couldn't find any.

**Comment by MasterJames** [comment]
Thank you for taking the time to help clarify some of the git issues better.

**Comment by tschw** [comment]
@tschw
> [... lots of things about Git ..]

@MasterJames 
> Thank you for taking the time to help clarify some of the git issues better.

Made it a Wiki page also covering more basic / other typical cases:

https://github.com/mrdoob/three.js/wiki/A-Contributor's-Guide-to-the-Git-Galaxy

**Comment by MasterJames** [comment]
Awesome! The new line glitch stuff is maybe moot to explain or ecen touch on with the MyBackup and reset concepts. Then again that might need it's own article. I suppose it's a windows thing. It's easy to want to use Gihub Desktop and/or my editors built it systems but it just adds more confusion.
I think there is another mention in a three contributors guide not to PR your main fork either, and the root of all the problems I had too. Thinking back to the moment I "JUMPED IN" (hastily), my regrettable assumption was you could make multiple forks as if they where the branches and that is important to git rit.

**Pull request closed without merging**